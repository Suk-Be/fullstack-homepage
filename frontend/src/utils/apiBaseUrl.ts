const apiBaseUrl = import.meta.env.VITE_API_BASE_URL;
const serverBaseUrl = import.meta.env.VITE_SERVER_BASE_URL;

export default apiBaseUrl;
export { serverBaseUrl };
