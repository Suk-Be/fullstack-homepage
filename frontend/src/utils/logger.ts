import { isAxiosError } from 'axios';

/**
 * Extract the HTTP status from an Axios error.
 */
export const getAxiosStatus = (error: unknown): number | null => {
    return isAxiosError(error) && error.response?.status ? error.response.status : null;
};

/**
 * Logs recoverable errors to console in dev or to backend in production.
 */

export const logRecoverableError = ({
  context,
  error,
  extra = {},
  mode = import.meta.env.MODE,
}: {
  context: string;
  error: unknown;
  extra?: Record<string, unknown>;
  mode?: string;
}) => {
  if (mode === 'test') return;

  if (mode === 'production') {
    if (typeof window !== 'undefined') {
      fetch('/api/log-client-error', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'X-Client-Secret': import.meta.env.VITE_CLIENT_LOG_SECRET,
        },
        body: JSON.stringify({
          context,
          error: isAxiosError(error) ? error.toJSON() : String(error),
          extra,
          timestamp: new Date().toISOString(),
        }),
      }).catch((err) => {
        console.error('[Logging failed]', err);
      });
    }
  } else {
    console.warn(`[DEV LOG] ${context}`, error, extra);
  }
};

export const logReduxState = (
  slice: 'login' | 'userGrid',
  state: any,
  mode: string = import.meta.env.MODE 
) => {
  if (mode === 'test' || mode === 'production') return;

  if (slice === 'login') {
    console.log('ðŸ”¹ login slice updated:', state);
  } else if (slice === 'userGrid') {
    console.log('ðŸ”¹ userGrid slice updated:', state);
  }
};


export const logRequestState = (
  requestType: 'initializeCookies' | 'initializeCookiesError' | 'requestMe' | 'requestMeError' | 'selectLoginState',
  responseType?: any,
  mode: string = import.meta.env.MODE 
) => {
  if (mode === 'test' || mode === 'production') return;

  if (requestType === 'initializeCookies') {
    console.log('[useAuthInit] Cookies initialized and throttled');
  } else if (requestType === 'initializeCookiesError') {
    console.log('[useAuthInit] Cookie init failed â†’ logout', responseType);
  } else if (requestType === 'requestMe') {
    console.log('[useAuthInit] requestMe() result:', responseType);
  } else if (requestType === 'requestMeError') {
    console.log('[useAuthInit] requestMe() failed â†’ logging out', responseType);
  } else if (requestType === 'selectLoginState') {
    console.log('[useAuthInit] Guard active â†’ already logged in (userId:', responseType, ')');
  }
};
