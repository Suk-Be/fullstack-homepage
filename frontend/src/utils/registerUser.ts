import LaravelApiClient from '../plugins/axios';

const isTestEnv = import.meta.env.MODE === 'test';

const registerUser = async ({
    logState,
    name,
    email,
    password,
    password_confirmation,
}: {
    logState: boolean;
    name: string;
    email: string;
    password: string;
    password_confirmation: string;
}) => {
    if (!isTestEnv) {
        await LaravelApiClient.post('/auth/spa/register', {
            name,
            email,
            password,
            password_confirmation,
        }).then((res) => {
            console.log('User', res.data);
        });

        if (logState) {
            // after await post, now get user data
            const { data } = await LaravelApiClient.get('/user');
            console.log('get User data, successful login', data);
        }
    }
};

export default registerUser;
