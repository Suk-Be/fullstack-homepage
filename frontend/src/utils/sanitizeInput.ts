import DOMPurify from 'dompurify';

export const sanitizeInput = (value: string): string => {
    // DOMPurify entfernt HTML, JS, Eventhandler etc.
    // und trimmt direkt
    return DOMPurify.sanitize(value, { ALLOWED_TAGS: [], ALLOWED_ATTR: [] }).trim();
};

interface SanitizeOptions {
    value: string;
    setValue: (val: string) => void;
    setError?: (msg: string) => void;
    friendlyMessage?: string;
}

/**
 * Sanitizes an input string and optionally notifies the user if the value was modified.
 *
 * @param options.value - The raw input value
 * @param options.setValue - State setter for the input value
 * @param options.setError - Optional state setter for error or info message
 * @param options.friendlyMessage - Optional message to show when sanitization occurs
 * @returns boolean - true if the value changed during sanitization
 */
export function sanitizeWithFeedback({
    value,
    setValue,
    setError,
    friendlyMessage = 'Some special characters were automatically removed. If this is okay with you, please confirm.',
}: SanitizeOptions): boolean {
    const sanitized = sanitizeInput(value);

    if (sanitized !== value) {
        setValue(sanitized);
        if (setError) {
            setError(friendlyMessage);
        }
        return true; // indicates a change happened
    }

    return false;
}
