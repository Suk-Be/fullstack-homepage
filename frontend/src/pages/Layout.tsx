import { ResponsiveContainer } from '@/components/ContainerElements';
import Footer from '@/components/footer';
import MenuNav from '@/components/header';
import { testId } from '@/utils/testId';
import CssBaseline from '@mui/material/CssBaseline';
import { useEffect } from 'react';
import { Toaster } from 'react-hot-toast';
import { Outlet } from 'react-router';

const Layout = () => {
    useEffect(() => {
        const banner = document.getElementById('cookie-banner');
        const accepted = localStorage.getItem('cookiesAccepted');
        if (!accepted && banner) {
            banner.style.display = 'block';
        }
    }, []);
    const acceptCookies = () => {
        localStorage.setItem('cookiesAccepted', 'true');
        const banner = document.getElementById('cookie-banner');
        if (banner) {
            banner.style.display = 'none';
        }
    };
    return (
        <>
            {/* css reset from material ui */}
            <CssBaseline />

            <ResponsiveContainer>
                <MenuNav />
                <div className="main-wrapper">
                    <main {...testId('main')}>
                        {/* container for content pages */}
                        <Outlet />
                        {/* container for toast notifications */}
                        <Toaster position="top-right" />
                    </main>
                    <Footer {...testId('footer')} />
                </div>
            </ResponsiveContainer>
            <aside
                id="cookie-banner"
                style={{
                    position: 'fixed',
                    bottom: 0,
                    left: 0,
                    right: 0,
                    background: '#222',
                    color: '#fff',
                    padding: '1rem',
                    textAlign: 'center',
                    zIndex: 9999,
                    display: 'none',
                }}
            >
                <span>
                    Diese Website verwendet Cookies, um die Funktionalit채t zu gew채hrleisten. Dar체ber
                    hinaus werden keine Daten erfasst. Mehr dazu in unserer{' '}
                    <a href="/datenschutz" style={{ color: '#4fc3f7' }}>
                        Datenschutzerkl채rung
                    </a>
                    .
                </span>
                <button
                    onClick={acceptCookies}
                    style={{
                        marginLeft: '1rem',
                        background: '#4fc3f7',
                        border: 'none',
                        color: '#000',
                        padding: '0.5rem 1rem',
                        cursor: 'pointer',
                    }}
                >
                    OK
                </button>
            </aside>
        </>
    );
};

export default Layout;
