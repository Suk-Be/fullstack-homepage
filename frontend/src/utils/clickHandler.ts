import apiBaseUrl from './apiBaseUrl';

export const handleSignInUp = (provider: string) => {
    window.location.href = `${apiBaseUrl}/auth/${provider}`;
};
