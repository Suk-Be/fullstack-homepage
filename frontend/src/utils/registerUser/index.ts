// export default registerUser;
import Cookies from 'js-cookie';
import apiBaseUrl from '../apiBaseUrl';
import logUserAfterRegistration from './logUserAfterRegistration';
import headers, { registerHeaders } from './requestHeaders';

const registerUser = async ({
    islog,
    name,
    email,
    password,
    password_confirmation,
}: {
    islog: boolean;
    name: string;
    email: string;
    password: string;
    password_confirmation: string;
}) => {
    let csrfToken: string | undefined;

    try {
        // Step 1: Get CSRF cookie
        await fetch(`${apiBaseUrl}/csrf-cookie`, {
            headers: headers,
            credentials: 'include',
        }).then(() => {
            csrfToken = Cookies.get('XSRF-TOKEN');
            if (!csrfToken) {
                throw new Error('CSRF token not found');
            }
        });

        // Step 2: Send register request
        if (csrfToken) {
            const response = await fetch(`${apiBaseUrl}/auth/spa/register`, {
                method: 'POST',
                credentials: 'include',
                headers: registerHeaders(csrfToken),
                body: JSON.stringify({
                    name: name,
                    email: email,
                    password: password,
                    password_confirmation: password_confirmation,
                }),
            });

            // Step 3: Return validation based on response status
            if (response.status === 422) {
                const errorData = await response.json();
                return { success: false, message: errorData.message || 'User already exists' };
            }
            if (response.status === 401) {
                const errorData = await response.json();
                return {
                    success: false,
                    message: errorData.message || 'Nicht autorisiert â€“ ggf. ausgeloggt',
                };
            }
            if (response.status === 419) {
                const errorData = await response.json();
                return { success: false, message: errorData.message || 'CSRF-Token abgelaufen' };
            }

            // Step 4: if isLog is set to true: the registered user data is fetched from backend
            await logUserAfterRegistration(islog, csrfToken);
        }
        // If everything is successful, return success which makes possible errors on email validation for already existing emails disappear
        return { success: true };
    } catch (error: any) {
        console.log('Status:', error);

        console.log('Status:', error);

        return { success: false, message: error };
    }
};

export default registerUser;
