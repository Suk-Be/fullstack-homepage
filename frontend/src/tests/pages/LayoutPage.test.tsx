import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import Layout from '../../pages/Layout';
import { navigateTo } from '../utils/testRenderUtils';

describe('LayoutPage', () => {
    beforeEach(() => {
        localStorage.clear(); // reset before each test
        document.body.innerHTML = ''; // ensure clean DOM
        vi.clearAllMocks();
    });

    afterEach(() => {
        localStorage.clear(); // cleanup
    });

    const renderUtil = (route: string) => {
        navigateTo({route: route});

        const header = screen.queryByTestId('header');
        const main = screen.queryByTestId('main');
        const footer = screen.queryByTestId('footer');

        const toaster = screen.queryByTestId('toaster');
        const routerOutlet = screen.queryByTestId('router-outlet');

        const logoLink = screen.queryByTestId('link-home-page');
        const impressumLink = screen.getByTestId('link-impressum-page');
        const datenschutzLink = screen.getByTestId('link-datenschutz-page');
        const avatarLink = screen.queryByTestId('button-open-menu');

        const user = userEvent.setup();

        return {
            header,
            main,
            footer,
            toaster,
            routerOutlet,
            logoLink,
            avatarLink,
            impressumLink,
            datenschutzLink,
            user,
        };
    };

    it('should render a main and a footer', () => {
            const { main, footer } = renderUtil('/');

            expect(main).toBeInTheDocument();
            expect(footer).toBeInTheDocument();
        },
    );

    // Do not remove Toaster or Outlet component
    it('includes MenuNav and Toaster in the JSX', () => {
      const tree = renderer.create(<Layout />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it.each([
        {
            page: 'HomePage',
            link: '/',
        },
        {
            page: 'ImprintPage',
            link: '/impressum',
        },
        {
            page: 'DatenschutzPage',
            link: '/datenschutz',
        },
    ])('shows cookie banner on $page if cookies not', async ({ link }) => {
        renderUtil(link);

        const banner = await screen.findByText(/diese website verwendet cookies/i);
        expect(banner).toBeVisible();
    });

    it.each([
        {
            page: 'HomePage',
            link: '/',
        },
        {
            page: 'ImprintPage',
            link: '/impressum',
        },
        {
            page: 'DatenschutzPage',
            link: '/datenschutz',
        },
    ])('hides cookie banner on $page and after accepting', async ({ link }) => {
        const { user } = renderUtil(link);

        const button = await screen.findByRole('button', { name: /ok/i });
        await user.click(button);

        await waitFor(() => {
            const banner = screen.queryByText(/diese website verwendet cookies/i);
            expect(banner).not.toBeVisible();
        });

        expect(localStorage.getItem('cookiesAccepted')).toBe('true');
    });

    it.each([
        {
            page: 'HomePage',
            link: '/',
        },
        {
            page: 'ImprintPage',
            link: '/impressum',
        },
        {
            page: 'DatenschutzPage',
            link: '/datenschutz',
        },
    ])('does not show banner if cookies were already accepted', ({ link }) => {
        localStorage.setItem('cookiesAccepted', 'true');
        renderUtil(link);

        const banner = screen.queryByText(/diese website verwendet cookies/i);
        expect(banner).not.toBeVisible();
    });
});
