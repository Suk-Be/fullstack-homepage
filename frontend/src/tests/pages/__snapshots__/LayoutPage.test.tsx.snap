// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LayoutPage > includes MenuNav and Toaster in the JSX 1`] = `null`;
